pkg/admission/karydia/karydia.go

```diff
@@ -48,129 +46,146 @@ func New(config *Config) (*KarydiaAdmission, error) {
-			if doesAutomountServiceAccountToken(&pod) {
-				validationErrors = append(validationErrors, "automount of service account not allowed")
-			}
-		} else if automountServiceAccountToken == "non-default" {
-			if doesAutomountServiceAccountToken(&pod) && pod.Spec.ServiceAccountName == "default" {
-				validationErrors = append(validationErrors, "automount of service account 'default' not allowed")
-			}
-		}
+	automountServiceAccountToken, annotated := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/automountServiceAccountToken"]
+	if annotated {
+		patches, validationErrors = admitServiceAccountToken(*pod, automountServiceAccountToken, patches, validationErrors)
 	}
 
-	seccompProfile, doCheck := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/seccompProfile"]
-	if doCheck {
-		seccompPod, ok := pod.ObjectMeta.Annotations["seccomp.security.alpha.kubernetes.io/pod"]
-		if !ok && mutationAllowed {
-			if len(pod.ObjectMeta.Annotations) == 0 {
-				// If no annotation object exists yet, we have
-				// to create it. Otherwise we will encounter
-				// the following error:
-				// 'jsonpatch add operation does not apply: doc is missing path: "/metadata/annotations..."'
-				patches = append(patches, fmt.Sprintf(`{"op": "add", "path": "/metadata/annotations", "value": {"%s": "%s"}}`, "seccomp.security.alpha.kubernetes.io/pod", seccompProfile))
-			} else {
-				patches = append(patches, fmt.Sprintf(`{"op": "add", "path": "/metadata/annotations/%s", "value": "%s"}`, "seccomp.security.alpha.kubernetes.io~1pod", seccompProfile))
-			}
-		} else if seccompPod != seccompProfile {
-			validationErrorMsg := fmt.Sprintf("seccomp profile ('seccomp.security.alpha.kubernetes.io/pod') must be '%s'", seccompProfile)
-			validationErrors = append(validationErrors, validationErrorMsg)
-		}
+	seccompProfile, annotated := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/seccompProfile"]
+	if annotated {
+		patches, validationErrors = admitSeccompProfile(*pod, seccompProfile, mutationAllowed, patches, validationErrors)
 	}
 
-	if len(validationErrors) > 0 {
-		return &v1beta1.AdmissionResponse{
-			Allowed: false,
-			Result: &metav1.Status{
-				Message: fmt.Sprintf("%+v", validationErrors),
-			},
+	return admitResponse(patches, validationErrors)
+}
+
+func admitServiceAccountToken(pod corev1.Pod, annotation string, patches []string, validationErrors []string) ([]string, []string) {
+	if annotation == "forbidden" {
+		// Validating webhook
+		if automountServiceAccountTokenUndefined(&pod) {
+			validationErrors = append(validationErrors, "automount of service account not allowed")
+		}
+	} else if annotation == "non-default" {
+		// Validating webhook
+		if automountServiceAccountTokenUndefined(&pod) && pod.Spec.ServiceAccountName == "default" {
+			validationErrors = append(validationErrors, "automount of service account 'default' not allowed")
+		}
+	} else if annotation == "remove-default" {
+		// Mutating webhook
```

Comment:
If karydia is used a validating webhook or mutating webhook or both is determined by how it's registered with the apiserver (`scripts/configure-karydia-webhook`). Internally, in the admission handlers, `mutationAllowed` was thought as a flag to distinguish the two cases and act accordingly:

* As "only" a validating webhook, karydia would either allow or deny the resource, depending on its configuration.
* As a mutating webhook, it could decide to patch a resource where applicable and only deny it if there are "unpatchable" violations of the configured rules.

In both cases, it should return meaningful and actionable error messages to the user, if any.

With this change, I can end up in a situation (installing karydia as a validating webhook only) where I set `karydia.gardener.cloud/automountServiceAccountToken=remove-default` and it doesn't have any effect. We will end up in the follow code path

```
	} else if annotation == "remove-default" {
		// Mutating webhook
```

karydia will return the patches to the apiserver, but they are ignored (rightfully, as it's only configured as a validating webhook here).

All this is not set in stone of course: you could for example say that karydia can or should be used only as a mutating webhook. Or you could make a clear difference between "rules" and "action items" and the latter can only be done when karydia is registered as mutating webhook.

One could argue it's a user error ("configuring `remove-default` makes no sense here") but I'm wary of the design decision and have the feeling that it make things more complex.

Reply:
> you could for example say that karydia can or should be used only as a mutating webhook.

Note that a "security add-on" like karydia must do validation though to enforce settings: https://github.com/karydia/karydia/issues/13